portable/GCC/ARM_CRx_MPU/port.c

```diff
@@ -0,0 +1,799 @@
+    ulRegionLength = ulRegionEnd - ulRegionStart;
+    ulRegionLength = prvGetMPURegionSizeSetting( ulRegionLength ) | portMPU_REGION_ENABLE;
+
+    prvMpuSetRegion(
+        portUNPRIVILEGED_FLASH_REGION,
+        ulRegionStart,
+        ulRegionLength,
+        portMPU_PRIV_RO_USER_RO_EXEC | portMPU_NORMAL_OIWTNOWA_SHARED
+    );
+
+    /* Privileged Read and Exec MPU Region for PRIVILEGED_FUNCTIONS. */
+    ulRegionStart = ( uint32_t ) __privileged_functions_start__;
+    ulRegionEnd = ( uint32_t ) __privileged_functions_end__;
+    ulRegionLength = ulRegionEnd - ulRegionStart;
+    ulRegionLength = prvGetMPURegionSizeSetting( ulRegionLength ) | portMPU_REGION_ENABLE;
+    prvMpuSetRegion(
+        portPRIVILEGED_FLASH_REGION,
+        ulRegionStart,
+        ulRegionLength,
+        portMPU_PRIV_RO_USER_NA_EXEC | portMPU_NORMAL_OIWTNOWA_SHARED
+    );
+
+    /* MPU Region for Peripheral Usage */
+    ulRegionStart = ( uint32_t ) __peripherals_start__;
+    ulRegionEnd = ( uint32_t ) __peripherals_end__;
+    ulRegionLength = ulRegionEnd - ulRegionStart;
+    ulRegionLength = prvGetMPURegionSizeSetting( ulRegionLength ) | portMPU_REGION_ENABLE;
+    prvMpuSetRegion(
+        portGENERAL_PERIPHERALS_REGION,
+        ulRegionStart,
+        ulRegionLength,
+        portMPU_PRIV_RW_USER_RW_NOEXEC | portMPU_DEVICE_NONSHAREABLE
+    );
+
+    /* Privileged Write and Read, Unprivileged Read, MPU Region for PRIVILEGED_DATA. */
+    ulRegionStart = ( uint32_t ) __privileged_data_start__;
+    ulRegionEnd = ( uint32_t ) __privileged_data_end__;
+    ulRegionLength = ulRegionEnd - ulRegionStart;
+    ulRegionLength = prvGetMPURegionSizeSetting( ulRegionLength ) | portMPU_REGION_ENABLE;
+    prvMpuSetRegion(
+        portPRIVILEGED_RAM_REGION,
+        ulRegionStart,
+        ulRegionLength,
+        portMPU_PRIV_RW_USER_RO_NOEXEC | portMPU_NORMAL_OIWTNOWA_SHARED
+    );
+
+    /* After setting default regions, enable the MPU */
+    prvMpuEnable();
+}
+
+/* ------------------------------------------------------------------------- */
+
+/** @brief Determine if a FreeRTOS Task has been granted access to a memory region
+ *
+ * @param xTaskMPURegion Pointer to a single set of MPU region registers.
+ * @param ulRegionStart Base address of the memory region access is being requested.
+ * @param ulRegionLength The length of the memory region that access is being requested.
+ * @param ulAccessRequested The type of access being requested, either read or write.
+ * @return BaseType_t pdTRUE if the task can access the region, pdFALSE otherwise
+ *
+ * @ingroup Task Context
+ * @ingroup MPU Control
+ */
+PRIVILEGED_FUNCTION static BaseType_t prvTaskCanAccessRegion(
+    const xMPU_REGION_REGISTERS * xTaskMPURegion,
+    const uint32_t ulRegionStart,
+    const uint32_t ulRegionLength,
+    const uint32_t ulAccessRequested
+)
+{
+    BaseType_t xAccessGranted;
+    uint32_t ulRegionEnd = ulRegionStart + ulRegionLength;
```

Comment:
Looks like we can potentially wrap around memory here, are there any checks at higher levels to ensure this condition wont happen? If not, might be best to add them here

Reply:
Wrap around check added by ensuring the end address is greater than the start address of the region